realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -145,8 +145,12 @@ public RealmObjectSchema addField(String fieldName, Class<?> fieldType, FieldAtt
                 containsAttribute(attributes, FieldAttribute.PRIMARY_KEY)) {
             nullable = false;
         }
-        long columnIndex = table.addColumn(metadata.realmType, fieldName, nullable);
+        // allow String & boxed variant type PrimaryKey to be null
+        if (containsAttribute(attributes, FieldAttribute.PRIMARY_KEY) && isNullablePrimaryKeyType(metadata)) {
```

Comment:
This is actually necessary. The reason this seem very redundant is because this particular line is a relaxation of #2003 @beeender has made. 

The lines from [144](https://github.com/realm/realm-java/blob/stk/null-primarykey/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java#L144) to [151](https://github.com/realm/realm-java/blob/stk/null-primarykey/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java#L151) basically indicates that if a column won't be nullable if it is marked as @Required or @PrimaryKey with primitive types.

I was thinking for a while to refactor them, then I realized that it would make us lose when and who has made changes especially when our final goal was to ultimately move toward @ObjectId.